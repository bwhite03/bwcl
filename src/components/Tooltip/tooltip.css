.bwcl-tooltip {
  position: relative;
}

.bwcl-tooltip-wrapper {
  position: relative;
  min-width: 120px;
  min-height: 120px;
}

.bwcl-tooltip-trigger {
  display: inline-block;
}

.bwcl-tooltip-bubble {
  min-width: 120px;
  position: absolute;
  z-index: 10;
  transition: opacity 750ms;
}

.bwcl-tooltip-bubble::after {
  content: '';
  position: absolute;
}

.bwcl-tooltip-top {
  bottom: 52%;
  left: 50%;
  padding-bottom: 9px;
  transform: translateX(-50%);
}

.bwcl-tooltip-top::after {
  border-left: 9px solid transparent;
  border-right: 9px solid transparent;
  border-top: 9px solid #fff;
  bottom: 0;
  left: 50%;
  transform: translateX(-50%);
}

.bwcl-tooltip-bottom {
  top: 52%;
  left: 50%;
  padding-top: 9px;
  transform: translateX(-50%);
}

.bwcl-tooltip-bottom::after {
  border-left: 9px solid transparent;
  border-right: 9px solid transparent;
  border-bottom: 9px solid #fff;
  left: 50%;
  top: 0;
  transform: translateX(-50%);
}

.bwcl-tooltip-left {
  top: 50%;
  right: 54%;
  padding-right: 9px;
  transform: translateY(-50%);
}

.bwcl-tooltip-left::after {
  border-left: 9px solid #fff;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  top: 50%;
  right: 0;
  transform: translateY(-50%);
}

.bwcl-tooltip-right {
  top: 50%;
  left: 54%;
  padding-left: 9px;
  transform: translateY(-50%);
}

.bwcl-tooltip-right::after {
  border-right: 9px solid #fff;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  top: 50%;
  left: 0;
  transform: translateY(-50%);
}

.bwcl-tooltip-message {
  background: #fff;
  border-radius: 3px;
  color: #000;
  font-size: 0.75rem;
  z-index: 9999;
}
