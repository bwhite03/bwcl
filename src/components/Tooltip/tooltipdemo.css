.tooltip-demo-container {
  display: flex;
}
