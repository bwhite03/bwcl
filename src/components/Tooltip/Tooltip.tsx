import React, { useState, useRef, useEffect } from 'react';
import './tooltip.css';

export type ToolTipPosition = 'top' | 'bottom' | 'left' | 'right';

export interface ToolTipProps {
  position: ToolTipPosition;
  children: React.ReactNode;
  message: string;
  style?: React.CSSProperties;
}

const ToolTip = (props: ToolTipProps) => {
  const [displayToolTip, setDisplayToolTip] = useState(false);

  let internalStyle = {
    backgroundColor: '#fff',
    color: '#000',
    padding: '5px',
  } as React.CSSProperties;

  const tooltipRef = useRef<HTMLSpanElement>(null);
  const triggerRef = useRef<HTMLSpanElement>(null);

  useEffect(() => {
    if (props.style) {
      internalStyle = props.style;
    }

    if (tooltipRef.current) {
      tooltipRef.current.addEventListener('mouseleave', hideToolTip);
    }

    if (triggerRef.current) {
      triggerRef.current.addEventListener('mouseover', showToolTip);
    }

    return () => {
      if (tooltipRef.current) {
        tooltipRef.current.removeEventListener('mouseleave', hideToolTip);
      }

      if (triggerRef.current) {
        triggerRef.current.removeEventListener('mouseover', showToolTip);
      }
    };
  }, []);

  const showToolTip = () => {
    setDisplayToolTip(true);
  };

  const hideToolTip = () => {
    setDisplayToolTip(false);
  };

  return (
    <span className="bwcl-tool-tip-wrapper" ref={tooltipRef}>
      {displayToolTip ? (
        <div>
          <div className={`bwcl-tooltip-bubble bwcl-tooltip-${props.position}`}>
            <div className="bwcl-tooltip-message" style={internalStyle}>
              {props.message}
            </div>
          </div>
        </div>
      ) : null}
      <span ref={triggerRef} className="bwcl-tooltip-trigger">
        {props.children}
      </span>
    </span>
  );
};

export default ToolTip;
