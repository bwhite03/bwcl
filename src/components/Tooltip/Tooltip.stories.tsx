import React from 'react';
import { Story, Meta } from '@storybook/react';
import ToolTip, { ToolTipProps } from './Tooltip';
import TooltipDemo, { ToolTipDemoProps } from './TooltipDemo';

export default {
  title: 'Tooltip',
  component: TooltipDemo,
  parameters: {
    layout: 'centered',
  },
} as Meta;

const Template: Story<ToolTipDemoProps> = (args) => <TooltipDemo {...args} />;
export const TooltipDemoExampleTop = Template.bind({});
TooltipDemoExampleTop.args = {
  position: 'top',
  message: 'top-demo',
};

export const TooltipDemoExampleBottom = Template.bind({});
TooltipDemoExampleBottom.args = {
  position: 'bottom',
  message: 'bottom-demo',
};

export const TooltipDemoExampleLeft = Template.bind({});
TooltipDemoExampleLeft.args = {
  position: 'left',
  message: 'left-demo',
};

export const TooltipDemoExampleRight = Template.bind({});
TooltipDemoExampleRight.args = {
  position: 'right',
  message: 'right-demo',
};
