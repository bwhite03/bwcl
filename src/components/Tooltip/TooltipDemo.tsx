import React from 'react';
import ToolTip from './Tooltip';
import { ToolTipPosition } from './Tooltip';
import './tooltipdemo.css';

export interface ToolTipDemoProps {
  position: ToolTipPosition;
  message: string;
}

const TooltipDemo = (props: ToolTipDemoProps) => {
  return (
    <div>
      <ToolTip
        position={props.position}
        message={props.message}
        style={{ marginTop: '20px' }}
      >
        <button>{props.message}</button>
      </ToolTip>
    </div>
  );
};

export default TooltipDemo;
