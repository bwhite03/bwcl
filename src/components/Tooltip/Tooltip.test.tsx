import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Tooltip, { ToolTipProps } from './Tooltip';

describe('Tooltip', () => {
  test('should render correctly', () => {
    const { container } = render(
      <Tooltip position="top" message="tooltip">
        <h2>test</h2>
      </Tooltip>
    );

    expect(container).toMatchSnapshot();
  });

  test('should render with style', () => {
    const { container } = render(
      <Tooltip
        position="top"
        message="tooltip"
        style={{ backgroundColor: 'red' }}
      >
        <h2>test</h2>
      </Tooltip>
    );

    expect(container).toMatchSnapshot();
  });

  test('should fire events', () => {
    const { container, getByRole, unmount } = render(
      <Tooltip position="top" message="tooltip">
        <h2>test</h2>
      </Tooltip>
    );

    const header = container.querySelector(
      '.bwcl-tooltip-trigger'
    ) as HTMLDivElement;
    fireEvent.mouseOver(header);
    fireEvent.mouseLeave(header);
  });

  test('should run useEffect cleanup', () => {
    const remover = jest
      .spyOn(React, 'useEffect')
      .mockImplementation((f) => f());
    const { unmount } = render(
      <Tooltip position="top" message="tooltip">
        <h2>Hello</h2>
      </Tooltip>
    );
    unmount();
  });
});
