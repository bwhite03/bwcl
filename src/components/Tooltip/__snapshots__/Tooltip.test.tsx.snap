// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should render correctly 1`] = `
<div>
  <span
    class="bwcl-tool-tip-wrapper"
  >
    <span
      class="bwcl-tooltip-trigger"
    >
      <h2>
        test
      </h2>
    </span>
  </span>
</div>
`;

exports[`Tooltip should render with style 1`] = `
<div>
  <span
    class="bwcl-tool-tip-wrapper"
  >
    <span
      class="bwcl-tooltip-trigger"
    >
      <h2>
        test
      </h2>
    </span>
  </span>
</div>
`;
